Add custom improvement proposal doc (#168)
evmos/docs: docs/develop/smart-contracts/custom-improvement-proposals.md

Within the evmOS framework, developers can modify any of the previous properties.

## Improvement Proposals

Improvement proposals are the approach used by evmOS and Ethereum to modify the
behavior of opcodes. They are composed by a function, which has access
to the jump table to apply specific changes to operations behavior, and a name.

In the context of Ethereum, these protocol changes are
named Ethereum Improvement Proposals (EIPs) and are identified by a unique ID.
For example, [EIP-1559](https://eips.ethereum.org/EIPS/eip-1559) is
used to introduce the base fee.

To allow any evmOS partner to define their own specific
improvements without overlapping with evmOS and Ethereum ones, each
proposal is identified by a string, that is composed of the chain name and a number. For
example, default evmOS improvements are associated with the string `evmos_XXXX`.
This allows each chain to define their improvements without having to worry
about existing or future ID clashes between different chains.
On top of that, the ability to start enumeration at 0 is better to handle for chain developers
and allows to have a better overview of the historical progress for each chain.

> 

Author: please, commit the change

# Custom Improvement Proposals

With the release [v19.0.0](https://github.com/evmos/evmos/releases/tag/v19.0.0)
a new feature called custom improvement proposals has been introduced in the
evmOS framework. Custom improvement proposals allow protocol developers to
modify the behavior of the EVM opcodes to tailor their functionalities to the
specific needs.

> 
> Improvement proposals are a way to introduce new standards, aimed to improve
> protocol functionalities. Changes proposed in this way can affect any aspect
> of the protocol but are mainly used to customize the behavior of
> smart contract execution.

Reviewer: I think this can actually be removed altogether, I think it's sufficiently explained in the paragraph above
```suggestion
```

# Custom Improvement Proposals

With the release [v19.0.0](https://github.com/evmos/evmos/releases/tag/v19.0.0)
a new feature called custom improvement proposals has been introduced in the
evmOS framework. Custom improvement proposals allow protocol developers to
modify the behavior of the EVM opcodes to tailor their functionalities to the
specific needs.

Improvement proposals are a way to introduce new standards, aimed to improve
protocol functionalities. Changes proposed in this way can affect any aspect

> of the protocol but are mainly used to customize the behavior of

Author: In the general sense of improvement proposal they are used to propose any change to the protocol. In the specific application of custom EIP via activator, they can only affect operations. If you think this is misleading I can reformulate!

> of the protocol but are mainly used to customize the behavior of

Reviewer: ahh okay gotcha - I think it would be more clear to just explain our specific case here :)

smart contract execution.

## Operations

Operations are the base components of the Ethereum Virtual Machine (EVM) which

> allow the execution of the smart contract logic. When a developer build a smart

Reviewer: ```suggestion
allow the execution of the smart contract logic. When a developer builds a smart
```

contract, the code written in Solidity, or Viper, is not directly interpretable
by the EVM. Before being able to execute the code in the blockchain, the

> contract has to be compiled via one of the available compilers, like **solc**. The

Reviewer: ```suggestion
contract has to be compiled via one of the available compilers, like [solc](https://docs.soliditylang.org/en/latest/using-the-compiler.html). The
```

> step of the compilation, convert the contract, written in a human readable
> language, into a sequence of operations that the virtual machine can interpret
> and execute to perform state transition or query the latest committed state.

Reviewer: ```suggestion
compilation converts the human readable contract code into a sequence of operations
that the virtual machine can interpret and execute
to perform state transitions or query the latest committed state.
```

> These operations are called, in the EVM context, **opcodes**, and are contained
> in a structure called **jump table**.

Reviewer: ```suggestion
These operations are called **opcodes**, and are contained
in a structure called [**jump table**](https://github.com/evmos/evmos/blob/v19.0.0/x/evm/core/vm/jump_table.go#L120-L1094).
```

> Some example of operations are the addition, defined by the opcode `ADD`, and
> the subtraction, defined by the opcode `SUB`.
> 

Reviewer: I think this can be removed
```suggestion
```

Each opcode is defined by specifying the logic that has to be executed when it
is called inside the EVM, its relationship with the memory, and the gas cost
associated with it. More specifically, an opcodes is completely defined by:

- `SetExecute`: update the execution logic for the opcode.

- `SetConstantGas`: update the value used for the constant gas cost.

- `SetDynamicGas`: update the function used to compute the dynamic gas cost.

- `SetMinStack`: update the minimum number of items in the stack required to
execute the operation.

- `SetMaxStack`: update the maximum number of items that will be in the stack
after executing the operation.

- `SetMemorySize`: the memory size required by the operation.

Within the evmOS framework, developers can modify any of the previous properties.

## Improvement Proposals

> Improvement proposal are the approach used by evmOS and Ethereum to modify the
> behavior of opcodes. They are composed by a function, which have the access
> to the jump table to apply specific changes to operations behavior, and a name.

Reviewer: ```suggestion
Improvement proposals are the approach used by evmOS and Ethereum to modify the
behavior of opcodes. They are composed by a function, which has access
to the jump table to apply specific changes to operations behavior, and a name.
```

> In the context of Ethereum, these protocol changes are
> named Ethereum Improvement Proposals (EIPs) and are associated to a numerical
> name. For example, the [EIP-1559](https://eips.ethereum.org/EIPS/eip-1559) is
> used to introduce the base fee.

Reviewer: ```suggestion
In the context of Ethereum, these protocol changes are
named Ethereum Improvement Proposals (EIPs) and are identified by a unique ID.
For example, [EIP-1559](https://eips.ethereum.org/EIPS/eip-1559) is
used to introduce the base fee.
```

> To allow any evmOS user to define their specific
> improvements without overlapping with evmOS and Ethereum default ones, each
> proposal is identified by a string composed by the chain name and a number. For

Reviewer: ```suggestion
To allow any evmOS partner to define their own specific
improvements without overlapping with evmOS and Ethereum ones, each
proposal is identified by a string, that is composed of the chain name and a number. For
```

example, default evmOS improvements are associated with the string `evmos_XXXX`.

> In this way, each chain can define their improvement without the risk of
> overwriting already present functions, and is free to start the numeration from
> 0 to have a more consistent history.

Reviewer: ```suggestion
This allows each chain to define their improvements without having to worry
about existing or future ID clashes between different chains.
On top of that, the ability to start enumeration at 0 is better to handle for chain developers
and allows to have a better overview of the historical progress for each chain.
```

> Below an example of how Evmos chain uses this functionalities to modify the
> behavior of the `CREATE` and `CREATE2` opcodes. First the modifier function has

Reviewer: ```suggestion
Below you will find an example of how the Evmos chain uses this functionality to modify the
behavior of the `CREATE` and `CREATE2` opcodes. First, the modifier function has
```

to be defined:

```go
// Enable0000 contains the logic to modify the CREATE and CREATE2 opcodes
// constant gas value.
func Enable0000(jt *vm.JumpTable) {
    multiplier := 10

	currentValCreate := jt[vm.CREATE].GetConstantGas()
	jt[vm.CREATE].SetConstantGas(currentValCreate * multiplier)

	currentValCreate2 := jt[vm.CREATE2].GetConstantGas()
	jt[vm.CREATE2].SetConstantGas(currentValCreate2 * multiplier)
}
```

Then, the function as to be associated with a name via a custom activator:

```go

> evmosActivators = map[int]func(*vm.JumpTable){
>     "evmos_0": eips.Enable0000,
> }

Reviewer: ```suggestion
evmosActivators = map[string]func(*vm.JumpTable){
    "evmos_0": eips.Enable0000,
}
```

```

## Activation of Improvement Proposals

Due to continuous changes in the users interaction with the protocol, and to
introduce a safety measure along with the freedom to customize the virtual

> machine behavior, custom improvement proposals are not active by default. The
> two just defined structures are used to define a modifier, and associate it with
> a namespace. The activation of selected improvement proposals is made via
> `x/evm` parameters. It is possible to activate specific improvement proposals in
> two ways:

Reviewer: ```suggestion
machine behavior, custom improvement proposals are not active by default. 
The activation of selected improvement proposals is controlled by the [EVM module's parameters](https://github.com/evmos/evmos/blob/main/proto/ethermint/evm/v1/evm.proto#L17-L18).
There are two ways of introducing the required parameter changes:
```

1. **Upgrade**: create a protocol upgrade handler which introduces the proposal name
in the active list.

> 2. **Governance**: create governance proposal to activate an improvement that is
> registered in the custom activator.

Reviewer: ```suggestion
2. **Governance**: create governance proposal to add an improvement proposal to the EVM module parameters.
```

> With this approach, it is given to developers the possibility to quickly react
> to security issues or market conditions, still keeping chain's participants in
> the loop.

Reviewer: ```suggestion
This approach gives developers the ability to react to security issues or market conditions,
while keeping the chain's participants in the loop.
```

## Additional Resources

1. [Evmos Custom EIPs](https://github.com/evmos/evmos/blob/main/app/eips/README.md):

> please refer to this document for a detailed description of how opcodes and
> custom improvement proposals has to be used in the evmOS framework.

Reviewer: ```suggestion
please refer to this document for a detailed description of how opcodes and
custom improvement proposals have to be used in the evmOS framework.
```

contract has to be compiled via one of the available compilers, like
[solc](https://docs.soliditylang.org/en/latest/using-the-compiler.html). The
compilation converts the human-readable contract code into a sequence of operations
that the virtual machine can interpret and execute
to perform state transitions or query the latest committed state.
These operations are called **opcodes**, and are contained
in a structure called [**jump table**](https://github.com/evmos/evmos/blob/v19.0.0/x/evm/core/vm/jump_table.go#L120-L1094).

Each opcode is defined by specifying the logic that has to be executed when it
is called inside the EVM, its relationship with the memory, and the gas cost
associated with it. More specifically, an opcode is completely defined by:

- `SetExecute` update the execution logic for the opcode.

- `SetConstantGas` update the value used for the constant gas cost.

- `SetDynamicGas` update the function used to compute the dynamic gas cost.

- `SetMinStack` update the minimum number of items in the stack required to
execute the operation.

- `SetMaxStack` update the maximum number of items that will be in the stack
after executing the operation.

> - `SetMemorySize` the memory size required by the operation.

Reviewer: how come you changed this? I think it was better before

Reviewer: if you're removing the `:` I think the text should be adjusted too, e.g. "`SetMemorySize` the memory size" is bad grammer. The `:` makes this correct

Author: Makes sense, I have been led astray from the rabbit...reverted

Reviewer: perfect, I think we're done then :)
